### PR TITLE
Rails 4.1.0.beta1

### PR DESCRIPTION
**Pull request by huoxito**
A couple changes to run on latest rails. Build still has a couple of failures, I believe they are related to the destroy action on dependent records or to the order the tests run, couldnt figure it out yet.

### CODE REVIEW

**Comment by huoxito** [comment]
all green on rails 4.1.0.beta1 :beers:

File: lib/paranoia.rb [hunk 1]
```diff
@@ -8,7 +8,11 @@ module Query
     def paranoid? ; true ; end
 
     def with_deleted
-      all.tap { |x| x.default_scoped = false }
+      if ActiveRecord::VERSION::STRING =~ /4\.1/
```
Comment (radar) [inline lib/paranoia.rb]: Please use `>` version comparsion here too.

File: lib/paranoia.rb [hunk 2]
```diff
@@ -47,6 +51,14 @@ def destroy
     run_callbacks(:destroy) { delete_or_soft_delete(true) }
   end
 
+  # As of Rails 4.1.0 +destroy!+ will only permanently destroy a record if it's
```
Comment (radar) [inline lib/paranoia.rb]: Please review the changes I did for #92 and edit this.

File: test/paranoia_test.rb [hunk 3]
```diff
@@ -1,13 +1,15 @@
-require 'test/unit'
 require 'active_record'
-require File.expand_path(File.dirname(__FILE__) + "/../lib/paranoia")
-
-DB_FILE = 'tmp/test_db'
 
-FileUtils.mkdir_p File.dirname(DB_FILE)
-FileUtils.rm_f DB_FILE
+test_framework = if ActiveRecord::VERSION::STRING =~ /4\.1/
```
Comment (radar) [inline test/paranoia_test.rb]: Please use `>` version comparison here.

File: test/paranoia_test.rb [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
+  require 'minitest/autorun'
+  MiniTest::Test
+else
+  require 'test/unit'
+  Test::Unit::TestCase
+end
+require File.expand_path(File.dirname(__FILE__) + "/../lib/paranoia")
 
-ActiveRecord::Base.establish_connection :adapter => 'sqlite3', :database => DB_FILE
+ActiveRecord::Base.establish_connection :adapter => 'sqlite3', database: ':memory:'
```
Comment (radar) [inline test/paranoia_test.rb:12]: :+1: Nice!

File: test/paranoia_test.rb [hunk 5]
```diff
@@ -263,12 +270,16 @@ def test_restore_behavior_for_callbacks
     assert model.instance_variable_get(:@restore_callback_called)
   end
 
-  def test_real_destroy
-    model = ParanoidModel.new
-    model.save
-    model.destroy!
+  # Only relevant to activerecord < 4.0 due to changes in
+  # Associations::HasManyAssociation see https://github.com/rails/rails/commit/5aab0c053832ded70a3a4b58cb97f8f8bba796ba
+  unless ActiveRecord::VERSION::STRING =~ /4\.1/
```
Comment (radar) [inline test/paranoia_test.rb]: Please use a `>` version comparsion here.

**Comment by huoxito** [comment]
@radar updated. As of the `destroy!` method it will only work as expected if users touch the deleted_at column before calling `destroy!`. Thats the only way I could think of the issue when destroy has_many associations due to changes in https://github.com/rails/rails/commit/5aab0c053832ded70a3a4b58cb97f8f8bba796ba. But maybe there's another way around it?

**Comment by radar** [comment]
@huoxito I don't understand your comment. the `destroy!` method is aliased to ActiveRecord's `destroy` (no bang) method, which will simply attempt to destroy the record... What did you mean with your comment?

**Comment by huoxito** [comment]
Sorry will try to explain again. The issue comes up when you try to _soft_delete_ a product for example (in spree context). It's supposed not to remove the product variants from the database but only touch its deleted_at columns. However if we don't override the `destroy!` implementation here it will remove those records. I think you can confirm that by removing the method and see that one test will fail.

I think the reason is that when destroying the has_many association records rails will now call, see https://github.com/rails/rails/blob/5aab0c053832ded70a3a4b58cb97f8f8bba796ba/activerecord/lib/active_record/associations/has_many_association.rb#L11, the activerecord/persistence `destroy!` method which in turn will call the activerecord/persistence `destroy` and so delete the record from database. hope that makes sense.

**Comment by radar** [comment]
I think you mean this line: https://github.com/rails/rails/blob/5aab0c053832ded70a3a4b58cb97f8f8bba796ba/activerecord/lib/active_record/associations/has_many_association.rb#L110-L111.

In that case, then yes that is very bad. Maybe we should call the method `i_am_really_sure_i_want_to_delete_this!` or something.

**Comment by huoxito** [comment]
@radar just rebased on top of rails4 branch again and added the recently created `really_destroy!` method to the tests. hopefully good to go now

**Comment by radar** [comment]
Added to rails4. Thanks @huoxito.

**Pull request closed without merging**